mixer/test/listbackend/list.go

```diff
@@ -0,0 +1,190 @@
+
+package listbackend
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net"
+	"sync"
+
+	"google.golang.org/grpc"
+
+	adptModel "istio.io/api/mixer/adapter/model/v1beta1"
+	"istio.io/istio/mixer/adapter/list"
+	"istio.io/istio/mixer/adapter/list/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/pkg/pool"
+	"istio.io/istio/mixer/pkg/runtime/handler"
+	"istio.io/istio/mixer/template/listentry"
+)
+
+type (
+	// Server is basic server interface
+	Server interface {
+		Addr() net.Addr
+		Close() error
+		Run()
+		Wait() error
+	}
+
+	// nosessionServer models no session adapter backend.
+	nosessionServer struct {
+		listener net.Listener
+		shutdown chan error
+		server   *grpc.Server
+
+		rawcfg      []byte
+		builder     adapter.HandlerBuilder
+		env         adapter.Env
+		builderLock sync.RWMutex
+
+		h listentry.Handler
+	}
+)
+
+var _ listentry.HandleListEntryServiceServer = &nosessionServer{}
+
+func (s *nosessionServer) getHandler(rawcfg []byte) (listentry.Handler, error) {
+	s.builderLock.RLock()
+	if 0 == bytes.Compare(rawcfg, s.rawcfg) {
+		h := s.h
+		s.builderLock.RUnlock()
+		return h, nil
+	}
+	s.builderLock.RUnlock()
+
+	// establish session
+	cfg := &config.Params{}
+	if err := cfg.Unmarshal(rawcfg); err != nil {
+		return nil, err
+	}
+
+	s.builderLock.Lock()
+	defer s.builderLock.Unlock()
+
+	if 0 == bytes.Compare(rawcfg, s.rawcfg) {
+		return s.h, nil
+	}
+
+	s.builder.SetAdapterConfig(cfg)
+	if ce := s.builder.Validate(); ce != nil {
+		return nil, ce
+	}
+
+	h, err := s.builder.Build(context.Background(), s.env)
+	if err != nil {
+		return nil, err
+	}
+	s.rawcfg = rawcfg
+	s.h = h.(listentry.Handler)
+	return s.h, err
+}
+
+func instance(in *listentry.InstanceMsg) *listentry.Instance {
+	return &listentry.Instance{
+		Name:  in.Name,
+		Value: in.Value,
+	}
+}
+
+// HandleListEntry records listrequest and responds with the programmed response
+func (s *nosessionServer) HandleListEntry(ctx context.Context, r *listentry.HandleListEntryRequest) (*adptModel.CheckResult, error) {
+	s.env.Logger().Infof("received '%v'", *r)
+	var h listentry.Handler
+	var err error
+	if r.AdapterConfig != nil {
+		h, err = s.getHandler(r.AdapterConfig.Value)
+	} else {
+		h, err = s.getHandler(nil)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	var cr adapter.CheckResult
+	if cr, err = h.HandleListEntry(ctx, instance(r.Instance)); err != nil {
+		return nil, err
+	}
+	s.env.Logger().Infof("response for '%v' is '%v'", *r, cr)
+	return &adptModel.CheckResult{Status: cr.Status, ValidUseCount: cr.ValidUseCount, ValidDuration: cr.ValidDuration}, nil
+}
+
+// Addr returns the listening address of the server
+func (s *nosessionServer) Addr() net.Addr {
+	return s.listener.Addr()
+}
+
+// Run starts the server run
+func (s *nosessionServer) Run() {
+	s.shutdown = make(chan error, 1)
+	go func() {
```

Comment:
As discussed in person, the goroutine here seems superfluous.

Reply:
WIll fix it in subsequent PR..need to get this for purpose of testing and demo